os/kernel/binary_manager/binary_manager_recovery.c

```diff
@@ -91,59 +97,94 @@ static void recovery_release_binary_sem(int binid)
 	irqrestore(flags);
 }
 
-static void recovery_exclude_scheduling_each(FAR struct tcb_s *tcb, FAR void *arg)
+/* Recover semaphores and watchdog timer resources */
+static void recovery_recover_tcb(struct tcb_s *tcb)
 {
-	int binid;
-	irqstate_t flags;
-
-	binid = (int)arg;
-	if (binid < 0) {
-		return;
-	}
+	sem_t *sem;
+	int state;
 
-	if (tcb->group->tg_loadtask == binid) {
+	state = tcb->task_state;
 
-		/* Recover semaphores, message queue, and watchdog timer resources.*/
-		task_recover(tcb);
+	if (tcb->waitdog) {
+		(void)wd_delete(tcb->waitdog);
+		tcb->waitdog = NULL;
+	}
 
-		flags = irqsave();
-		/* Remove the TCB from the task list associated with the state */
-		dq_rem((FAR dq_entry_t *)tcb, (dq_queue_t *)g_tasklisttable[tcb->task_state].list);
-		sched_addblocked(tcb, TSTATE_TASK_INACTIVE);
-		irqrestore(flags);
-		bmllvdbg("Remove pid %d from task list\n", tcb->pid);
+	if (state == TSTATE_WAIT_SEM) {
+		sem = tcb->waitsem;
+		DEBUGASSERT(sem != NULL && sem->semcount < 0);
```

Comment:
Yes. I think it has some problems in this case because sem, mq can be kernel resources.
So if these conditions are not matched, system will be stopped.